pkgs/development/compilers/cudatoolkit/extension.nix

```diff
@@ -38,10 +38,42 @@ final: prev: let
 
   cudaFlags = final.callPackage ./flags.nix {};
 
+  # Internal hook, used by cudatoolkit and cuda redist packages
+  # to accommodate automatic CUDAToolkit_ROOT construction
+  markForCudatoolkitRootHook = (final.callPackage
+    ({ makeSetupHook }:
+      makeSetupHook
+        { name = "mark-for-cudatoolkit-root-hook"; }
+        ./hooks/mark-for-cudatoolkit-root-hook.sh)
+    { });
+
+  # Normally propagated by cuda_nvcc or cudatoolkit through their depsHostHostPropagated
+  setupCudaHook = (final.callPackage
+    ({ makeSetupHook, backendStdenv }:
+      makeSetupHook
+        {
+          name = "setup-cuda-hook";
+
+          substitutions.ccRoot = "${backendStdenv.cc}";
+
+          # Required in addition to ccRoot as otherwise bin/gcc is looked up
+          # when building CMakeCUDACompilerId.cu
+          substitutions.ccFullPath = "${backendStdenv.cc}/bin/${backendStdenv.cc.targetPrefix}c++";
+
+          # Required by cmake's enable_language(CUDA) to build a test program
+          substitutions.cudartRoot = prev.lib.getDev (final.cuda_cudart or final.cudatoolkit);
```

Comment:
I removed the setup hook from nvcc and merged it here instead, so we have one big `setupCudaHook`. This is still more complex than I hoped for, and we'll have to revisit the paths soon when implementing cross-compilation

Reply:
Whelp, I broke it. Cudatoolkit uses this hook, so for cudaPackages_10_1 it's infinite recursion

Reply:
My feeling is that `deps<foo><bar>Propagated` is just a bad idea

Reply:
Reverted to when `cuda_nvcc` had its own hook for cudart. I'll open an issue for removing the hook `depsTargetTargetPropagated` (and maybe removing `backendStdenv`) in favour of explicit

```nix
mkDerivation {
  # ...
  nativeBuildInputs = [
    # ...
    cudaPackages.setupCudaHook
    cudaPackages.cuda_nvcc
  ];
  # ...
}
```


The reason to not do that in this PR is that this would require touching many downstream packages, making the diff and the blame messy, and pinging more people.